theano/tensor/nnet/nnet.py

```diff
@@ -130,329 +130,496 @@ def c_code_template(dtype):
+    E_axis = 'invalid axis'
     __props__ = ()
 
-    def make_node(self, x):
+    def make_node(self, x, axis=-1):
         x = tensor.as_tensor_variable(x)
-        if x.type.ndim not in (1, 2) \
-                or x.type.dtype not in tensor.float_dtypes:
-            raise ValueError('x must be 1-d or 2-d tensor of floats. Got %s' %
-                             x.type)
+        axis = tensor.as_tensor_variable(axis)
+        # TODO : Delete this and modify the test accordly
         if x.ndim == 1:
             x = tensor.shape_padleft(x, n_ones=1)
-
-        return Apply(self, [x], [x.type()])
+        inputs = [x, axis]
+        return Apply(self, inputs, [x.type()])
 
     def perform(self, node, input_storage, output_storage):
-        x, = input_storage
-        e_x = np.exp(x - x.max(axis=1)[:, None])
-        sm = e_x / e_x.sum(axis=1)[:, None]
+        x, axes = input_storage
+        # Apply softmax on the specified dimension
+        e_x = np.exp(x - x.max(axis=axes, keepdims=True))
+        sm = e_x / e_x.sum(axis=axes, keepdims=True)
         output_storage[0][0] = sm
 
     def L_op(self, inp, outputs, grads):
-        x, = inp
+        x, axes = inp
         g_sm, = grads
-        return [softmax_grad(g_sm, outputs[0])]
+        axes_grad = theano.gradient.grad_undefined(self, 1, axes)
+        return [abstract_softmax_grad(g_sm, outputs[0], axes), axes_grad]
 
     def R_op(self, inputs, eval_points):
-        # I think the Jacobian is symmetric so the R_op
-        # is the same as the grad
-        if None in eval_points:
+        # The Jacobian is symmetric so the R_op is the same as the grad
+        if eval_points[0] is None:
             return [None]
-        return self.L_op(inputs, [self(*inputs)], eval_points)
+        return self.L_op(inputs, [self(*inputs)], [eval_points[0]])
 
     def infer_shape(self, node, shape):
-        return shape
+        return [shape[0]]
+
+abstract_softmax_op = Abstract_softmax()
+
+
+class Softmax(Abstract_softmax):
+    """
+    Softmax activation function
+    :math:`\\varphi(\\mathbf{x})_j =
+    \\frac{e^{\mathbf{x}_j}}{\sum_{k=1}^K e^{\mathbf{x}_k}}`
+    where :math:`K` is the total number of neurons in the layer. This
+    activation function gets applied on the selected axis (Default axis is the last).
+    This function get applied when axis is a scalar and uses a fast C optimization.
+    """
+
+    nin = 2  # tensor, axis
+    nout = 1
+    __props__ = ()
+
+    def make_node(self, x, axis=-1):
+        x = tensor.as_tensor_variable(x)
+        axis = tensor.as_tensor_variable(axis)
+        if x.ndim == 1:
+            x = tensor.shape_padleft(x, n_ones=1)
+        # Check if axis is an integer
+        if axis.type.dtype not in tensor.int_dtypes:
+            raise ValueError('axis must be an integer. Got ', axis.type)
+        # Check if axis is a scalar
+        try:
+            opt.get_scalar_constant_value(axis)
+        except tensor.NotScalarConstantError:
```

Comment:
check_and_normalize_axis return a list of sorted integers. This Softmax op with the C code optimization corresponds to the old op that manages only the case where we specify **only one axis**. So, if check_and_normalize_axis return (2,3) and this op is called instead of the Abstract_Softmax, it will fail because the C code doesn't manage more than one axis. I think we should really keep it because in case where there are some issue in the optimization it will be much easier to detect the case where a wrong op is called.